### ISSUE TITLE
Generate named nested types

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi-terraform#163]
Today, we only generate named types for the outermost resource arguments, we do not generate named types for nested structure within that.  This means that there is no way in TypeScript to write functions which return a piece of that nested structure without manually duplicating part of 

It should be possible to generate named types for some or all of this nested structure as well.  This could be either an auotmatic thing, or something that individual resources can opt-in to.  

There will be some challenges in figuring out where these types should be shared between multiple resources (using a single name), but it still be useful even if that sharing doesn't exist - due to structural typing in TypeScript allowing the types to be used interchangeably.

**Comment by lukehoban** [comment]
From @brandonbloom:

> Seems like this might already be tracked somewhere, but I couldn't find an issue. Apologies if I missed something.

> The TypeScript declarations are super helpful for discovering what options are available on various resources. However, there are many deeply nested configuration objects that are built almost entirely of `pulumi.Input`, anonymous objects, and primitives. If you make a small error in a nested attribute, `tsc` produces pretty large and obnoxious error messages that are difficult to read. The entire root configuration object gets a squiggly red underline!

> I've found that a great way to deal with this problem is to build up larger objects incrementally and assign explicit types to the intermediate values. Although restructuring syntax helps, this is a little more verbose, but well worth it, as it significantly narrows the scope of compiler errors.

> Sometimes, you can use fancy TypeScript features like indexed types to get at a useful anonymous type for this purpose. For example:

> ```typescript
> let defaultCacheBehavior: aws.cloudfront.DistributionArgs['defaultCacheBehavior'] = { ...
> ```

> But that's pretty kludgy and sometimes not possible, for example when something is wrapped in `pulumi.Input<...>[]`.

> Would be nice if all of these various properties had usefully named types. For this example, I'd love to write:

> ```typescript
> let defaultCacheBehavior: aws.cloudfront.DefaultCacheBehavior = { ...
> ```

> Generally, there are sensible names for these things in the CloudFormation documentation. Again, for example: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudfront-distribution-defaultcachebehavior.html

**Comment by joeduffy** [comment]
I've taken a few whacks at this and (annoyingly) actually had a working change but seem to have lost it somewhere in my stash stack.

My approach was simply to auto-name the types, however I love @brandonbloom's idea of allowing explicit names. I wonder if we should auto-name them by default and then permit overriding the names using overlays like we do for lots of other automatic code-gen behavior. (The one downside to this is that renaming is a breaking change -- however the advantage is that we don't need to manually come up with names for the thousands of types that could benefit from this.)

For example, take the Brandon's example of the `aws.cloudfront.Distribution`'s (monstrous) `cacheBehaviors` property:

```typescript
class Distribution {
    cacheBehaviors: pulumi.Output<{ allowedMethods: string[], cachedMethods: string[], compress?: boolean, defaultTtl?: number, fieldLevelEncryptionId?:     string, forwardedValues: { cookies: { forward: string, whitelistedNames?: string[] }, headers?: string[], queryString: boolean, queryStringCacheKeys?: string[] }, lambda    FunctionAssociations?: { eventType: string, lambdaArn: string }[], maxTtl?: number, minTtl?: number, pathPattern: string, smoothStreaming?: boolean, targetOriginId: stri    ng, trustedSigners?: string[], viewerProtocolPolicy: string }[] | undefined>;
...
}
```

The auto-name scheme I devised would have led to something like:

```typescript
class Distribution {
    cacheBehaviors: pulumi.Output<DistributionCacheBehavior[] | undefined>;
...
}

interface DistributionCacheBehavior {
    allowedMethods: string[];
    cachedMethods: string[];
    compress?: boolean;
    defaultTtl?: number;
    fieldLevelEncryptionId?: string;
    forwardedValues: DistributionCacheBehaviorForwardedValues;
    lambdaFunctionAssociations?: DistributionCacheBehaviorLambdaFunctionAssociation[];
    maxTtl?: number;
    minTtl?: number;
    pathPattern: string;
    smoothStreaming?: boolean;
    targetOriginId: string;
    trustedSigners?: string[];
    viewerProtocolPolicy: string;
}

interface DistributionCacheBehaviorForwardedValues {
    cookies: DistributionCacheBehaviorForwardValuesCookies;
    headers?: string[];
    queryString: boolean;
    queryStringCacheKeys?: string[];
}

interface DistributionCacheBehaviorForwardValuesCookies {
    forward: string;
    whitelistedNames?: string[];
}

interface DistributionCacheBehaviorLambdaFunctionAssociation {
    eventType: string;
    lambdaArn: string;
}
```


This example brings up an interesting other hitch, which is that `aws.cloudfront.Distribution` _also_ has a `defaultCacheBehavior` field which is almost identical, except that it's not an array. It would be ideal if those could be unified somehow, but given the auto-name scheme above, they wouldn't be. It's possible there's some structural approach here, but if we can manually unify those that matter with overlays -- as we would for this case -- that could be a nice sweet spot.

**Comment by pgavlin** [comment]
@lukehoban @justinvp I think this is done, yes?

**Comment by justinvp** [comment]
Yes, this was fixed with #456.
